pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -583,11 +587,28 @@ func buildGatewayNetworkFiltersFromTLSRoutes(node *model.Proxy, env *model.Envir
 	return filterChains
 }
 
-func pickMatchingGatewayHosts(gatewayServerHosts map[model.Hostname]bool, virtualServiceHosts []string) map[string]model.Hostname {
+// Select the virtualService's hosts that match the ones specified in the gateway server's hosts
+// based on the wildcard hostname match and the namespace match
+func pickMatchingGatewayHosts(gatewayServerHosts map[model.Hostname]bool, virtualService model.Config) map[string]model.Hostname {
```

Comment:
"gatewaysServerHosts includes all normalized hosts in the gateway we are picking config for. Method is called for each virtual service in the mesh "

Also add a comment on what is cached and what not - we need to start making it clear in the comments on which methods are O(n^2) but cached, etc. I think this method in particular is called 
VirtualserviceCount * GatewayInstanceCount - for example 5000 services * 10 gateways ?

Reply:
Yes. Number of VirtualServices bound to the gateway * Number of servers in the merged Gateway. So, if you have 5K virtualServices bound to the same gateway workload, and the gateway config has 10 Servers (or 10 gateways with one server each), then function is effectively called 50K times. And unfortunately, this function is not cacheable because this gets invoked when there is a change in the config (virtual service or gateway). 
It can certainly be cached once per Gateway service (e.g. istio-ingressgateawy) and reused across all gateway instances [100s of replicas].